File: src/echoes_mcp/database/schemas.py
"""Pydantic schemas for LanceDB tables."""

from typing import Literal

from pydantic import BaseModel, Field

EntityType = Literal["CHARACTER", "LOCATION", "EVENT", "OBJECT", "EMOTION"]

RelationType = Literal[
    # Character relations
    "LOVES",
    "HATES",
    "KNOWS",
    "RELATED_TO",
    "FRIENDS_WITH",
    "ENEMIES_WITH",
    # Spatial relations
    "LOCATED_IN",
    "LIVES_IN",
    "TRAVELS_TO",
    # Temporal relations
    "HAPPENS_BEFORE",
    "HAPPENS_AFTER",
    "CAUSES",
    # Object relations
    "OWNS",
    "USES",
    "SEEKS",
]


class ChapterRecord(BaseModel):
    """Schema for chapters.lance table."""

    # Identification
    id: str = Field(description="Unique ID: arc:episode:chapter")
    file_path: str = Field(description="Relative path to markdown file")
    file_hash: str = Field(description="SHA256 hash for change detection")

    # Hierarchy
    arc: str
    episode: int
    chapter: int

    # Metadata from frontmatter
    pov: str = Field(description="Point of view character")
    title: str
    location: str | None = None
    date: str | None = Field(default=None, description="Narrative date")

    # Content
    content: str = Field(description="Clean text without frontmatter/markdown")
    excerpt: str | None = Field(default=None, description="Short summary")

    # Statistics
    word_count: int
    char_count: int
    paragraph_count: int

    # RAG
    vector: list[float] = Field(description="768-dim embedding")
    entities: list[str] = Field(default_factory=list, description="Entity IDs")

    # Metadata
    indexed_at: int = Field(description="Unix timestamp")


class EntityRecord(BaseModel):
    """Schema for entities.lance table."""

    id: str = Field(description="Unique ID: type:name")
    name: str
    type: EntityType
    description: str
    aliases: list[str] = Field(default_factory=list)

    # RAG
    vector: list[float] = Field(description="Embedding of name+description")

    # References
    chapters: list[str] = Field(default_factory=list, description="Chapter IDs")
    chapter_count: int = 0
    first_appearance: str | None = None

    # Metadata
    indexed_at: int


class RelationRecord(BaseModel):
    """Schema for relations.lance table."""

    id: str = Field(description="Unique ID: source:type:target")
    source_entity: str = Field(description="Source entity ID")
    target_entity: str = Field(description="Target entity ID")
    type: str = Field(description="Relation type")
    description: str
    weight: float = Field(ge=0.0, le=1.0, description="Importance/frequency")

    # References
    chapters: list[str] = Field(default_factory=list)

    # Metadata
    indexed_at: int

File: lib/database/schemas.ts
import { Field, FixedSizeList, Float32, Float64, Int32, List, Schema, Utf8 } from 'apache-arrow';

// Entity and Relation type constants
export const ENTITY_TYPES = ['CHARACTER', 'LOCATION', 'EVENT', 'OBJECT', 'EMOTION'] as const;
export type EntityType = (typeof ENTITY_TYPES)[number];

export const RELATION_TYPES = [
  'LOVES',
  'HATES',
  'KNOWS',
  'RELATED_TO',
  'FRIENDS_WITH',
  'ENEMIES_WITH',
  'LOCATED_IN',
  'LIVES_IN',
  'TRAVELS_TO',
  'HAPPENS_BEFORE',
  'HAPPENS_AFTER',
  'CAUSES',
  'OWNS',
  'USES',
  'SEEKS',
] as const;
export type RelationType = (typeof RELATION_TYPES)[number];

// Dynamic Arrow Schema factories
export function createChapterSchema(embeddingDim: number): Schema {
  return new Schema([
    new Field('id', new Utf8(), false),
    new Field('file_path', new Utf8(), false),
    new Field('file_hash', new Utf8(), false),
    new Field('arc', new Utf8(), false),
    new Field('episode', new Int32(), false),
    new Field('chapter', new Int32(), false),
    new Field('pov', new Utf8(), false),
    new Field('title', new Utf8(), false),
    new Field('location', new Utf8(), false),
    new Field('date', new Utf8(), false),
    new Field('content', new Utf8(), false),
    new Field('summary', new Utf8(), false),
    new Field('word_count', new Int32(), false),
    new Field('char_count', new Int32(), false),
    new Field('paragraph_count', new Int32(), false),
    new Field('vector', new FixedSizeList(embeddingDim, new Field('item', new Float32())), false),
    new Field('entities', new List(new Field('item', new Utf8())), false),
    new Field('indexed_at', new Float64(), false),
  ]);
}

export function createEntitySchema(embeddingDim: number): Schema {
  return new Schema([
    new Field('id', new Utf8(), false),
    new Field('arc', new Utf8(), false),
    new Field('name', new Utf8(), false),
    new Field('type', new Utf8(), false),
    new Field('description', new Utf8(), false),
    new Field('aliases', new List(new Field('item', new Utf8())), false),
    new Field('vector', new FixedSizeList(embeddingDim, new Field('item', new Float32())), false),
    new Field('chapters', new List(new Field('item', new Utf8())), false),
    new Field('chapter_count', new Int32(), false),
    new Field('first_appearance', new Utf8(), false),
    new Field('indexed_at', new Float64(), false),
  ]);
}

export const RelationSchema = new Schema([
  new Field('id', new Utf8(), false),
  new Field('arc', new Utf8(), false),
  new Field('source_entity', new Utf8(), false),
  new Field('target_entity', new Utf8(), false),
  new Field('type', new Utf8(), false),
  new Field('description', new Utf8(), false),
  new Field('weight', new Float64(), false),
  new Field('chapters', new List(new Field('item', new Utf8())), false),
  new Field('indexed_at', new Float64(), false),
]);

// TypeScript interfaces

// Base chapter data from scanner (without RAG fields)
export interface ScannedChapter {
  id: string;
  file_path: string;
  file_hash: string;
  arc: string;
  episode: number;
  chapter: number;
  pov: string;
  title: string;
  location: string;
  date: string;
  content: string;
  summary: string;
  word_count: number;
  char_count: number;
  paragraph_count: number;
}

// Full chapter record with RAG fields for database
export interface ChapterRecord extends ScannedChapter {
  vector: number[];
  entities: string[];
  indexed_at: number;
}

export interface EntityRecord {
  id: string;
  arc: string;
  name: string;
  type: EntityType;
  description: string;
  aliases: string[];
  vector: number[];
  chapters: string[];
  chapter_count: number;
  first_appearance: string;
  indexed_at: number;
}

export interface RelationRecord {
  id: string;
  arc: string;
  source_entity: string;
  target_entity: string;
  type: RelationType;
  description: string;
  weight: number;
  chapters: string[];
  indexed_at: number;
}
